pkg/asset/store.go

```diff
@@ -11,50 +12,86 @@ import (
 type Store interface {
 	// Fetch retrieves the state of the given asset, generating it and its
 	// dependencies if necessary.
-	Fetch(Asset) error
+	// It returns dirty if the asset or any of its parents is loaded from on-disk files.
+	Fetch(Asset) (dirty bool, err error)
 }
 
 // StoreImpl is the implementation of Store.
 type StoreImpl struct {
-	assets map[reflect.Type]Asset
+	// OnDiskFiles stores the any existing files under the cluster directory,
+	// so that each asset can load and reuse them.
+	OnDiskFiles map[string][]byte
+	assets      map[reflect.Type]Asset
 }
 
 // Fetch retrieves the state of the given asset, generating it and its
 // dependencies if necessary.
-func (s *StoreImpl) Fetch(asset Asset) error {
+func (s *StoreImpl) Fetch(asset Asset) (dirty bool, err error) {
 	return s.fetch(asset, "")
 }
 
-func (s *StoreImpl) fetch(asset Asset, indent string) error {
+func (s *StoreImpl) fetch(asset Asset, indent string) (dirty bool, err error) {
 	logrus.Debugf("%sFetching %s...", indent, asset.Name())
+
+	// Return immediately if the asset is found in the cache,
+	// this is because we are doinga depth-first-search, it's guaranteed
+	// that we always fetch the parent before children, so we don't need
+	// to worry about invalidating anything in the cache.
 	storedAsset, ok := s.assets[reflect.TypeOf(asset)]
 	if ok {
 		logrus.Debugf("%sFound %s...", indent, asset.Name())
 		reflect.ValueOf(asset).Elem().Set(reflect.ValueOf(storedAsset).Elem())
-		return nil
+		return false, nil
```

Comment:
This shouldn't always be false for dirty. The store needs to retain whether the asset is dirty because it may be a parent of multiple other assets.